### PR TITLE
Partial Approval, ApplePay and AndroidPay certification tests

### PR DESCRIPTION
**Pull request by alenad**
@pierre

### CODE REVIEW

File: db/ddl.sql [hunk 1]
```diff
@@ -78,5 +78,6 @@ CREATE TABLE `litle_responses` (
   `updated_at` datetime NOT NULL,
   `kb_account_id` varchar(255) DEFAULT NULL,
   `kb_tenant_id` varchar(255) DEFAULT NULL,
+  `approved_amount` int(11) DEFAULT NULL,
```
Comment (pierre) [inline db/ddl.sql]: What is this for / what does it represent?

If we need it, can you:

* Prefix it with `params_` (convention)
* Place it next to the other `params_` columns (somewhere near lines 63-69 above)
* Provide the [migration script](http://docs.killbill.io/latest/database_migrations.html#_ruby_plugins)

Also, is an `approved_currency` returned as well? If so, let's add it (could matter for cross-border transactions).

File: lib/litle.rb [hunk 2]
```diff
@@ -94,6 +97,15 @@ def parse(kind, xml)
 
         parsed
       end
+
+      def success_from(kind, parsed)
+        if kind == :registerToken
+          %w(000 801 802).include?(parsed[:response])
+        else
+          %w(000 010).include?(parsed[:response])
```
Comment (pierre) [inline lib/litle.rb]: It looks like the main difference with upstream is the support of `010`? Could you document it?

File: spec/litle/remote/certification/certification_spec.rb [hunk 3]
```diff
@@ -303,6 +303,94 @@
...
+                                         :cc_exp_month => '09',
+                                         :cc_exp_year => '2021',
+                                         :allow_partial_auth => true,
+                                     },
+                                     false)
+
+    assertions = {
+        :success => true,
+        :message => 'Partially Approved',
+        :approvedAmount => 32000,
```
Comment (pierre) [inline spec/litle/remote/certification/certification_spec.rb]: Nit: should it be `approved_amount` instead?

File: lib/litle/api.rb [hunk 4]
```diff
@@ -70,11 +70,22 @@ def credit_payment(kb_account_id, kb_payment_id, kb_payment_transaction_id, kb_p
...
-        # Pass extra parameters for the gateway here
-        options = {}
+        gateway_call_proc = Proc.new do |gateway, linked_transaction, payment_source, amount_in_cents, options|
+          gateway.refund(amount_in_cents, linked_transaction.txn_id, options)
+        end
 
-        properties = merge_properties(properties, options)
-        super(kb_account_id, kb_payment_id, kb_payment_transaction_id, kb_payment_method_id, amount, currency, properties, context)
+        linked_transaction_proc = Proc.new do |amount_in_cents, options|
+          # we need the id of either a capture or sale transaction
```
Comment (pierre) [inline lib/litle/api.rb]: Interesting - do you know if that has changed? I'm surprised it hasn't been caught earlier.

Is there a specific certification test we could add for regression?

**Review by pierre — CHANGES_REQUESTED** [review]

File: db/ddl.sql (hunk 1 above)
Comment (alenad) [inline db/ddl.sql]: This is to approve partial authorizations, if a full amount authorization cannot be secured.

**Pull request closed without merging**